src/core/auth/qgsauthmanager.cpp

```diff
@@ -497,11 +497,13 @@ bool QgsAuthManager::setMasterPassword( bool verify )
   if ( mMasterPass.isEmpty() )
   {
     QgsDebugMsg( QStringLiteral( "Master password is not yet set by user" ) );
+    locker.unlock();
```

Comment:
I think masterPasswordInput() must be protected from concurrent accesses. Maybe the best solution would be to move the lockers down one level into the methods but I understand that it would be a major rewrite of the locking system in the manager.

Reply:
@elpaso , a lock could take place inside qgscredential, but I can't see a way it can be done in the authentication manager itself that'd work across threads.

In any case, I wouldn't undertake that unless there's an actual bug to fix. ATM, this PR fixes a QGIS freeze/crash and doesn't come with regressions on travis / azure nor my workflows here.

Reply:
@elpaso , or, there could be a master password mutex, independent of the overall authentication manager mutex, to prevent multiple master password credential requests.